### ISSUE TITLE
Design dataset loading utilities

### ISSUE THREAD

**Issue by LukeWood** [keras-team/keras-cv#325]
Recently I went through the process to load a bounding box dataset from a CSV file and a directory of images.  I think there are some nice abstractions we could offer to help in this process.

I’ll post more details after thinking more about API design.  Feel free to leave comments of ideas here.

**Comment by bhack** [comment]
Like https://github.com/tensorflow/datasets/blob/master/tensorflow_datasets/object_detection/kitti.py ?

**Comment by bhack** [comment]
We we start to talk about dataset adapters I think a lot will have to be thought about whether, with the limited resources available, we want to compete with TFDS in the ecosystem or cooperate.

**Comment by LukeWood** [comment]
> When we start to talk about datasets adapters I think a lot will have to be thought about whether, with the limited resources available, we want to compete with TFDS in the ecosystem or cooperate.

The goal of KerasCV is to offer the most cohesive, pleasant user experience possible for training state of the art computer vision models.  TensorFlow datasets is a full featured powerful API that is quite nice to use.  To me, doesn’t mean that there’s no space for a layer of abstraction that fits together coherently with KerasCV.  

I think this comment represents my thoughts on all of the discussions we have relating to this topic.  This only represents my own opinion on the matter, and as always happy to discuss more

**Comment by LukeWood** [comment]
Regarding TFDS: I’m not envisioning utilities for specific datasets, but rather utilities to load custom datasets that abide to specific formats with tf.data.

**Comment by bhack** [comment]
Let see what do you want to introduce pratically here when you have detailed this idea.

Keras has a BDFL governance so I just share  an opinion.

My opinion Is that if you start to overfit coherence you could go to duplicate/overlap more then cooperate as you need to stay inside your perimeter to maintain the API design control.

**Comment by kartik4949** [comment]
@LukeWood 
I did similar work on this 
https://github.com/kartik4949/TensorPipe

Features:

* Automatic loading of bounding box data, mask data, key points data, classification data.
* State of the art augmentation 
* prefetch, optimised pipeline, caching, etc optimisation inbuilt pipeline
* custom injection of lumpy function in your tf dataset pipeline.
* extreme configuration and flexibility.

**Comment by kartik4949** [comment]
@LukeWood 
I did similar work on this 
https://github.com/kartik4949/TensorPipe

Features:

* Automatic loading of bounding box data, mask data, key points data, classification data.
* State of the art augmentation 
* prefetch, optimised pipeline, caching, etc optimisation inbuilt pipeline
* custom injection of lumpy function in your tf dataset pipeline.
* extreme configuration and flexibility.

**Comment by kartik4949** [comment]
basically an advance Keras image generator but for mask, bboxes, etc.

**Comment by LukeWood** [comment]
Thanks @kartik4949 for the reference!  I think we can separate out data loading and augmentation to make things customizable, but we should provide easy to use documentation on how to combine them.

**Comment by kartik4949** [comment]
@LukeWood thanks for taking a look at the library, seems like ill work on the library again, and will refactor it a bit :)

and YES
data loading should be separated out from augmentations.

I feel like, its the api design that will make this stand out!

My thoughts:

```python
import keras_cv

loader = keras_cv.DataLoader(path=/path/to/train/data, type="classification")
data_iterable = loader.load()
augmented_data_iterable = keras_cv.apply_augmentation_pipeline(aug_pipeline, data_iterable)
```

**Comment by LukeWood** [comment]
I'm still not fully sure what we want to offer here; this will require a lot of consideration with respect to:
- TPU distribution
- tfrecord preparation
- etc

Basically, we want to offer something useful and unique; not reinvent the wheel.  Its not clear to me exactly what this will look like, but my hope is that as we go through the process of creating some training templates this becomes clearer.

**Comment by kartik4949** [comment]
> I'm still not fully sure what we want to offer here; this will require a lot of consideration with respect to:
> 
> * TPU distribution
> * tfrecord preparation
> * etc
> 
> Basically, we want to offer something useful and unique; not reinvent the wheel. Its not clear to me exactly what this will look like, but my hope is that as we go through the process of creating some training templates this becomes clearer.

make sense!

**Comment by LukeWood** [comment]
Done for now - maybe we revisit someday but for now let’s keep our offerings narrow, and only offer loaders for datasets that are useful for tutorials
